File: advent_of_code_2017/day3/fiboweird.py
from collections import defaultdict

def dimension(n):
    m = 0
    while n > (2 * m + 1)**2:
        m += 1
    return 2 * m + 1


def elm(row, col):
    max_rowcol = max(abs(row), abs(col))
    dim = max_rowcol * 2 + 1
    if row == -max_rowcol:
        return dim**2 - (max_rowcol - col)
    elif col == -max_rowcol:
        return dim**2 - dim - (max_rowcol + row) + 1
    elif row == max_rowcol:
        return dim**2 - 2 * dim + 2 - (max_rowcol + col)
    return dim**2 - 3 * dim + 3 - (max_rowcol - row)


def pos(idx):
    dim = dimension(idx)
    max_rowcol = (dim - 1) / 2
    bottom_left = dim**2 - dim + 1
    top_left = dim**2 - 2 * dim + 2
    top_right = dim**2 - 3 * dim + 3
    if idx >= bottom_left:
        return (-max_rowcol, max_rowcol - (dim**2 - idx))
    elif idx >= top_left:
        return (-max_rowcol + (bottom_left - idx), -max_rowcol)
    elif idx >= top_right:
        return (max_rowcol, -max_rowcol + (top_left - idx))
    return (max_rowcol - (top_right - idx), max_rowcol)


def adjacents(position):
    (a, b) = position
    return [(a + i, b + j) for i in range(-1, 2) for j in range(-1, 2)]


def walk(destination):
    matrix = defaultdict(lambda: 0)
    matrix[(0, 0)] = 1
    n = 1
    while matrix[pos(n)] <= destination:
        n += 1
        n_pos = pos(n)
        elements = [matrix[p] for p in adjacents(pos(n))]
        result = sum(elements)
        matrix[pos(n)] = result
    return matrix[pos(n)]

# print(pos(22))
# print(elm(*pos(22)))
# print([pos(n) for n in [2,3,4,6,8]])
# print([elm(*pos(n)) for n in [2,3,4,6,8]])
print(walk(368078))

File: advent_of_code_2017/day3/fiboweird.hs
import qualified Data.Map as Map

main :: IO ()
main = do
  putStrLn (show (adjacents (0, 0)))
  putStrLn (show (dimension 24))
  putStrLn ("part2: " ++ (show (successor 368078)))

elm :: (Int, Int) -> Int
elm (row, col) = do
  let max_rowcol = max (abs row) (abs col)
  let dim = max_rowcol * 2 + 1
  if row == -max_rowcol
    then dim ^ 2 - (max_rowcol - col)
    else if col == -max_rowcol
           then dim ^ 2 - dim + 1 - (max_rowcol + row)
           else if row == max_rowcol
                  then dim ^ 2 - 2 * dim + 2 - (max_rowcol + col)
                  else dim ^ 2 - 3 * dim + 3 - (max_rowcol - row)

dimension :: Int -> Int
dimension n = head ([2 * m + 1 | m <- [1 ..], n <= (2 * m + 1) ^ 2])

pos :: Int -> (Int, Int)
pos idx = do
  let dim = dimension idx
  let max_rowcol = div (dim - 1) 2
  let bottom_left = dim ^ 2 - dim + 1
  let top_left = dim ^ 2 - 2 * dim + 2
  let top_right = dim ^ 2 - 3 * dim + 3
  if idx >= bottom_left
    then (-max_rowcol, max_rowcol - (dim ^ 2 - idx))
    else if idx >= top_left
           then (-max_rowcol + (bottom_left - idx), -max_rowcol)
           else if idx >= top_right
                  then (max_rowcol, -max_rowcol + (top_left - idx))
                  else (max_rowcol - (top_right - idx), max_rowcol)

successor :: Int -> Int
successor n = do
  let matrix = Map.fromList [((0, 0), 1)]
  do_successor matrix n 1

default0 :: Maybe Int -> Int
default0 maybe_int = maybe 0 id maybe_int

do_successor :: Map.Map (Int, Int) Int -> Int -> Int -> Int
do_successor matrix destination n = do
  let maybe_val_at_pos = Map.lookup (pos n) matrix
  let val_at_pos = default0 maybe_val_at_pos
  if val_at_pos > destination
    then val_at_pos
    else do
      let n' = n + 1
      let adjacent_pos = adjacents (pos n')
      let elements = [Map.lookup x matrix | x <- adjacent_pos]
      let elem_values = [default0 x | x <- elements]
      let new_elem = sum elem_values
      let matrix' = Map.insert (pos n') new_elem matrix
      do_successor matrix' destination n'

adjacents :: (Int, Int) -> [(Int, Int)]
adjacents (a, b) = [(x, y) | x <- [a - 1, a, a + 1], y <- [b - 1, b, b + 1]]
